course5/homework06/ustring.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <iostream>
+#include <stdexcept>
+#include <limits>
+#include <memory>
+#include <cassert>
+
+namespace Course {
+
+class String
+{
+public:
+    static constexpr std::size_t npos = -1;
+
+    String() = default;
+
+    String(const char *str) { *this = str; }
+    String(const char *str, size_t length);
+    String(String const &str) { *this = str; }
+    String(String &&rhs);
+
+    String& operator=(const char *str);
+    String& operator=(String const &rhs);
+    String& operator=(String &&rhs);
+
+    // Element access
+    char& operator[](std::size_t pos) { return data()[pos]; }
+
+    const char& operator[](std::size_t pos) const { return data()[pos]; }
+
+    char& at(std::size_t pos)
+    {
+        if (pos >= mLength)
+            throw std::out_of_range("String::at");
+        return (*this)[pos];
+    }
+
+    const char& at(std::size_t pos) const
+    {
+        if (pos >= mLength)
+            throw std::out_of_range("String::at");
+        return (*this)[pos];
+    }
+
+    char& front() { return (*this)[0]; }
+    const char& front() const { return (*this)[0];  }
+
+    char& back() { return (*this)[mLength-1]; }
+    const char& back() const { return (*this)[mLength-1]; }
+
+    char* data()
+    {
+        if (length() == 0)
+            return emptyString;
```

Comment:
Since C++11 std::string::data() must return a pointer to single null character when length is zero: http://en.cppreference.com/w/cpp/string/basic_string/data
Yes, it's a lie, because it's write access to something, but there is really no place to write to. I wonder how the standard library implements this.